core/chain_manager.go

```diff
@@ -139,14 +148,16 @@ func (bc *ChainManager) SetHead(head *types.Block) {
 	defer bc.mu.Unlock()
 
 	for block := bc.currentBlock; block != nil && block.Hash() != head.Hash(); block = bc.GetBlock(block.ParentHash()) {
-		bc.removeBlock(block)
+		DeleteBlock(bc.chainDb, block.Hash())
 	}
+	bc.headerCache.Purge()
```

Comment:
*[Note, `SetHead` is not used for doing chain reogrs. It is used by the console `debug.setHead(blocknum)` command to rewind the chain to some previous block.]*

This purging is in `SetHead`, which essentially rewinds the chain to a previous state (deleting everything in between). But by removing data from the database, we need to make sure that everything rolled back is also removed from the caches, otherwise the system will see inconsistent state.